File: src/main/java/com/skidders/sigma/module/Module.java
package com.skidders.sigma.module;

import com.skidders.sigma.module.settings.Setting;
import com.skidders.sigma.util.client.interfaces.IMinecraft;
import com.skidders.sigma.util.client.interfaces.ISubscriber;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Setter;

import java.util.ArrayList;
import java.util.List;

@RequiredArgsConstructor
@Getter
@Setter
public class Module implements IMinecraft, ISubscriber {

    private final String name, desc;
    private final Category category;
    private boolean enabled;
    private int key;

    public Module(String name, String desc, Category category, int key) {
        this(name, desc, category);
        this.key = key;
    }

    public void onEnable() {}
    public void onDisable() {}
    public void onInit() {}

    public void setEnabled(boolean enabled) {
        this.enabled = enabled;

        if (enabled) {
            bus.register(this);
            onEnable();
        } else {
            onDisable();
            bus.unregister(this);
        }
    }

    public List<Setting<?>> settings = new ArrayList<>();

    public void registerSetting(Setting<?> setting) {
        if (!settings.contains(setting)) {
            settings.add(setting);
        } else {
            throw new IllegalArgumentException("Attempted to add an duplicate setting.");
        }
    }

    public Setting<?> getSettingByName(String input) {
        return settings.stream().filter(s -> s.name.equalsIgnoreCase(input)).findFirst().orElse(null);
    }

}

File: src/main/kotlin/com/skidders/sigma/module/Module.kt
package com.skidders.sigma.module

import com.skidders.SigmaReborn
import com.skidders.sigma.module.settings.Setting
import com.skidders.sigma.utils.IMinecraft

import java.util.ArrayList

open class Module : IMinecraft {

    var enabled: Boolean
    var name: String
    var desc: String
    val category: Category
    var key: Int? = null

    constructor(name: String, desc: String, category: Category) {
        this.name = name
        this.desc = desc
        this.category = category
        this.enabled = false
    }

    constructor(name: String, desc: String, category: Category, key: Int) : this(name, desc, category) {
        this.key = key
    }

    open fun onEnable() {}
    open fun onDisable() {}

    fun setEnabled(enabled: Boolean) {
        this.enabled = enabled

        if (enabled) {
            SigmaReborn.EVENT_BUS.register(this)
            onEnable()
        } else {
            onDisable()
            SigmaReborn.EVENT_BUS.unregister(this)
        }
    }

    val settings: ArrayList<Setting<*>> = ArrayList<Setting<*>>()

    fun registerSetting(setting: Setting<*>) {
        if (!settings.contains(setting)) {
            settings.add(setting)
        } else {
            throw IllegalArgumentException("Attempted to add an duplicate setting.")
        }
    }

    fun getSettingByName(input: String): Setting<*> {
        return settings.stream().filter { it.name.equals(input, ignoreCase = true) }.findFirst()
            .orElse(null)
    }

}
